### ISSUE TITLE
JSON Response Issue When Integrating Into Yesod

### ISSUE THREAD

**Issue by tuxagon** [morpheusgraphql/morpheus-graphql#396]
This may not be an issue and maybe it's actually an issue with a different library, but I'm coming here first for potential direction since it's specifically around usage of this library.

I have created a simple, isolated example based on a yesod template 

https://github.com/tuxagon/haskell-graphql-yesod

Within that repo, the second commit shows how GraphQL was set up and the only file, I believe, that matters is `src/Handler/Graphql.hs`. 

Everything is _mostly_ hooked up correctly. The only issue is that the response is 
```
{
  "tag": "Data",
  "contents": {
    "deity": {
      "name": "Morpheus"
    }
  }
}
```

instead of 

```
{
  "data": {
    "deity": {
      "name": "Morpheus"
    }
  }
}
```

I am not quite sure why scotty is providing the correct JSON with

```
api :: B.ByteString -> IO B.ByteString
api = interpreter rootResolver

main :: IO ()
main =
  scotty 4000 $ post "/api" $ raw =<< (liftIO . api =<< body)
```

but the following in Yesod is giving me that incorrect JSON above

```
graphqlApi :: GQLRequest -> IO GQLResponse
graphqlApi =
  interpreter rootResolver

postGraphqlR :: Handler Value
postGraphqlR = do
  body <- requireCheckJsonBody :: Handler GQLRequest
  result <- (liftIO . graphqlApi) body
  returnJson result
```

Everything else is the same between the 2 frameworks and essentially just the README's examples.

I suspect the different contexts might be at play with scotty's `ActionM ByteString` vs yesod's `Handler GQLResponse` in these particular examples, but if so, I'm not quite sure where to go from there.

Any help or guidance would be appreciated and I hope that I've provided enough context.

**Comment by tuxagon** [comment]
I think I found an approach that works here. There was a lot of digging between this library, aeson, and yesod to figure out how I could resolve the types in a way that made both morpheus and yesod happy. I couldn't change anything to use just `ByteString` because yesod needs something that implements the `ToJSON` type class and `Text` just doesn't work when I pass it to `graphqlApi`. So those restrictions ruled out using aeson's `encode` and manually using `toJSON`.

Instead, I decided to see what I can change the resulting type from the function to. It uses a type class called `ToTypedContent` for return types. Since `Value` was used in this example and is from a non-yesod library, I knew I should be able to return something that implements the `ToJSON` type class that isn't part of yesod. Digging a little deeper, I found `JSONResponse a` and decided to give it a whirl before implementing a new `ToTypedContent` instance.

The resulting function is 
```
postGraphqlR :: Handler (JSONResponse GQLResponse)
postGraphqlR = do
  body <- requireCheckJsonBody :: Handler GQLRequest
  result <- (liftIO . graphqlApi) body
  return $ JSONResponse result
```

It seems to work. Whether this is preferred over a new type class instance is something I don't have an answer for since I am still fairly new to yesod.

I hope others find this useful. I'm going to close it.
